### ISSUE TITLE
Failed to install CoverM

### ISSUE THREAD

**Issue by timbalam** [wwood/CoverM#2]
Hello!

I haven't been successful at building or installing CoverM. I have tried using Rust on Windows targeting the Gnu ABI, and on Ubuntu installation in the Windows Subsystem for Linux.  When building with 'cargo run' both times I have had issues compiling rust-htslib, with the error:

```
error: failed to run custom build command for `rust-htslib v0.19.1`
process didn't exit successfully: `/home/timothy/git/CoverM/target/debug/build/rust-htslib-fa9ca6cca9341e9a/build-script-build` (exit code: 101)
--- stderr
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }', libcore/result.rs:945:5
note: Run with `RUST_BACKTRACE=1` for a backtrace. 
```

(The backtrace is not helpful). I thought it might be due to htslib not supporting windows very well, but when the Linux Subsystem also didn't work, I thought I'd let you know in case there was something else going on. 

I also tried installing with `cargo install coverm` to see if that would work, but it doesn't seem to be available. I get this error:
```
$ cargo install coverm
    Updating registry `https://github.com/rust-lang/crates.io-index`
error: could not find `coverm` in registry `https://github.com/rust-lang/crates.io-index`
```

Cheers,
Tim

**Comment by wwood** [comment]
Hi Tim,
Ah yes, it seems I never got around to pushing it to crates.io, so the README about cargo install coverm is not right. I'll push release something properly soon once #1 is merged.

With the rust-htslib thing, You could try updating the 0.19 to something newer and see if that helps. Presumably the error is caused by some missing file specified in the https://github.com/rust-bio/rust-htslib/blob/master/build.rs although without more details I'm not sure I can help. Really, this is an issue for https://github.com/rust-bio/rust-htslib/issues

Can't you just not use Windows :P

**Comment by rhysnewell** [comment]
Hi Tim,
There is some strange interactions happening with rust-htslib and libclang currently. Current temporary work around  is to specify your cargo build with the path to libclang e.g.

`LIBCLANG_PATH=/usr/lib/llvm-6.0/lib cargo build`

Going to work out why this is ASAP.

Also in regards to install on windows... Currently I don't think it's possible as the rust-htslib module is not compatible with windows. Neither is base htslib either, or samtools. But, you can get hts-lib and rust-htslib to build on the windows linux subsystem by using the above method involving LIBCLANG_PATH.

**Comment by timbalam** [comment]
Thanks Rhys and Ben,

I guess I mainly want to know whether you'd expect rust-htslib etc. to
install  on the windows linux subsystem.

Maybe I can install guix on there...

**Comment by wwood** [comment]
Guix might work, but maybe another option is that we've managed to compile it staticly. Would that help?

**Comment by timbalam** [comment]
How does one compile it statically? Is there a cargo option?

**Comment by rhysnewell** [comment]
`cargo build --target x86_64-unknown-linux-musl`
 will compile statically. Make sure you have musl gcc installed and add the x86_64-unknown-linux-musl target to your rustup e.g. 
`rustup target add x86_64-unknown-linux-musl`

Also, you have to be on the head branch of rust-htslib, not their latest release version as the current release doesn't support musl.

**Comment by wwood** [comment]
Since we've already done it though, testing is just a simple matter of running the compiled version, which I've shared here. Note that this is a development version of coverm.
https://cloudstor.aarnet.edu.au/plus/s/cvDt57VhWkqdWxb

Works?

**Comment by timbalam** [comment]
Sorry, I didn't get around to trying the download. It appears to be down
at the moment. Is it a temporary link?

I went ahead and tried to statically compile rust-htslib. It didn't work
at first, but it gave a more helpful error message.

I needed the development headers for the rust-htslib dependencies, zlib,
etc.

Installing those meant that I could statically build rust-htslib. I
couldn't statically build CoverM though, but I could build it regularly.

I guess I just needed the rust-htslib dependencies? Maybe I'll delete
everything and try to repeat the regular build with the rust-htslib
dependencies.

**Comment by timbalam** [comment]
I tried rebuilding with the rust-htslib dependencies. I also needed to
install gcc, but now it seems to be working.

Hooray!

It's weird that you'd need the dependencies to use rust-htslib as a
library though, right?

**Comment by timbalam** [comment]
After rereading the rust-htslib readme, maybe it's not weird.

**Comment by wwood** [comment]
Hi Tim,

Strange about the link - it is working for me.

You have to install the non-rust dependencies of rust-htslib because rust works a little differently. When you compile coverm, you are compiling not just coverm but also its entire software stack - it isn't like python where e.g. once you pip install graftm, pip install singlem uses that previous install.

Anyway, glad you got there in the end. Let us know if you find any bugs or have any improvements.

ben
